rust/kernel/types.rs

```diff
@@ -37,9 +36,22 @@ impl Mode {
 /// in kernel data structures, for example, an implementation of [`FileOperations`] in `struct
 /// file::private_data`.
 pub trait PointerWrapper {
+    /// Type of values borrowed between calls to [`PointerWrapper::into_pointer`] and
+    /// [`PointerWrapper::from_pointer`].
+    type Borrowed: Deref;
+
     /// Returns the raw pointer.
     fn into_pointer(self) -> *const c_types::c_void;
 
+    /// Returns a borrowed value.
+    ///
+    /// # Safety
+    ///
+    /// `ptr` must have been returned by a previous call to [`PointerWrapper::into_pointer`].
+    /// Additionally, [`PointerWrapper::from_pointer`] can only be called after *all* values
+    /// returned by [`PointerWrapper::borrow`] have been dropped.
+    unsafe fn borrow(ptr: *const c_types::c_void) -> Self::Borrowed;
```

Comment:
That's not true. `*const` is just a pointer and it does not impose any restriction unless it is dereferenced. `*const T` and `*mut T` is less different between `&` and `&mut`. In fact, the only real differences are:
* `*const T` is covariant and `*mut T` is invariant
* `*const T` dereferences to `&T` and `*mut T` dereferences to `&mut T`

Reply:
```rust
unsafe fn borrow(ptr: *const c_types::c_void) -> Self::Borrowed;
```
This function takes a `*const` pointer, and returns an object which allows the data behind the `*const` pointer to be modified through interior mutability. I guess it doesn't really matter, because
> `ptr` must have been returned by a previous call to [`PointerWrapper::into_pointer`]

But isn't it potentially confusing that mutable data is passed around via a `*const` pointer?

Reply:
Why is interior mutability a problem here? Interior mutability works through `&` so it is natural to have `*const`.

Reply:
Does interior mutability work in C through a `const` pointer? E.g. if I have a
```C
struct foo {
    struct mutex m;
    int i;
};
```
and I have a `const` pointer to it:
```C
const foo *f;
```
Can I then use the `const` pointer above to, say, lock the mutex and increment `i`?
I am guessing that it does work in Rust via a shared reference:
```rust
struct foo(Mutex<i32>);
fn change_it(f: &foo) {
    *f.lock().unwrap() += 1;
}
```

Reply:
You can't in C, but you can in Rust or C++. Rust's `*const T` is more like C++'s `const T*` rather than C's; `UnsafeCell` can be seen as C++'s `mutable`.

Reply:
I see, that makes sense.

The `PointerWrapper` abstraction is designed to produce a pointer which is passed (assigned) to C. That pointer is a `*const`, so it can be assigned to a `const void *` pointer in a C structure. Which implies that the data being pointed at, must not/cannot be written to. But then in `borrow()` we take that `const void *` pointer, and create a type which can mutate the data !
```rust
unsafe fn borrow(ptr: *const c_types::c_void) -> Self::Borrowed;
```
Doesn't that sound confusing?

Reply:
> Just out of curiosity, is there a Rust pointer or reference that absolutely guarantees that the data is immutable?

No.

> For example, how would one point at data in ROM, or a `ro` section? The kind of memory that generates an exception if someone tries to modify it.

Rust uses a compiler-internal `Freeze` trait (https://doc.rust-lang.org/nightly/src/core/marker.rs.html#712-724) to determine if something could be placed in rodata. If the type of a static implements `Freeze`, it is placed in rodata automatically.

Reply:
In rust `*const T` and `*mut T` are identical from the perspective of the abstract machine. They both allow mutating the pointee if the pointer was derived from a mutable reference. If the pointer was derived from an immutable reference, writing is only allowed when there is explicit interior mutability through `UnsafeCell`. The only difference between `*const T` and `*mut T` is in the type system.

Reply:
> Is it correct to use `*const c_types::c_void` for the pointer? Or should it be `*mut c_types::c_void` instead?

`PointerWrapper` mimics the [`into_raw`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.from_raw) and [`from_raw`](https://doc.rust-lang.org/std/sync/struct.Arc.html#method.into_raw) functions from `Arc` but loses information about the underlying type (`T` vs `void`) because it's stored by C code. This function is an addition for something to be used between those calls, so it must have the same input type as `from_pointer`.

So while the discussion about data stored in `.ro` is important (and we should check that immutable tables of functions that we generate in cost functions are indeed placed in `.ro`), I don't think it will result in us having to change this type. And even if it does, it will likely be a bulk change in a bunch of places, which is something better done in a different PR.

Reply:
Now that I realize `*const T` allows interior mutability, and is quite different from C's `const T *`, I agree that we are using the correct type here.